src/model/sample.rs

```diff
@@ -36,40 +36,73 @@ pub fn prob_read_base(read_base: u8, ref_base: u8, base_qual: u8) -> LogProb {
+                &Cigar::SoftClip(_) => {
+                    c = i;
+                    break;
+                },
+                &Cigar::Del(_) => panic!("Cigar String: Unexpected Deletion operation found before any reference match."),
+                &Cigar::Back(_) => panic!("Back: Unsupported Cigar operation (not in SAMv1 spec)."),
+                &Cigar::Pad(_) => panic!("Pad: Unsupported Cigar operation (unclear SAMv1 spec)."),
+                &Cigar::RefSkip(_) => panic!("Cigar String: Unexpected Reference Skip operation found before any reference match."),
+                &Cigar::HardClip(_) => ()
+            }
+
+        }
+        while rpos <= vpos {
+            match &cigar[c] {
                 // potential SNV evidence
-                &Cigar::Match(l) | &Cigar::Diff(l) if contains_start(pos, l) => {
+                &Cigar::Match(l) | &Cigar::Diff(l) | &Cigar::Equal(l) if contains_vpos(rpos, l) => {
+                    qpos += vpos - rpos; // difference between vpos and first position of current cigar operation
                     let read_base = record.seq()[qpos as usize];
                     let base_qual = record.qual()[qpos as usize];
                     let prob_alt = prob_read_base(read_base, base, base_qual);
-                    let prob_ref = prob_read_base(read_base, ref_seq[start as usize], base_qual);
-                    return (prob_ref, prob_alt);
+                    let prob_ref = prob_read_base(read_base, ref_seq[vpos as usize], base_qual);
+                    return Some( (prob_ref, prob_alt) );
                 },
                 // for others, just increase pos and qpos as needed
                 &Cigar::Match(l)   |
-                &Cigar::RefSkip(l) |
-                &Cigar::Equal(l)   |
                 &Cigar::Diff(l)    |
+                &Cigar::Equal(l)   => {
+                    rpos += l;
+                    qpos += l;
+                    c += 1;
+                },
+                &Cigar::SoftClip(l) |
                 &Cigar::Ins(l)  => {
-                    pos += l as i32;
                     qpos += l;
+                    c += 1;
                 },
-                &Cigar::Del(l)  => pos += l as i32,
-                &Cigar::Back(l) => pos -= l as i32,
-                _ => ()
+                &Cigar::RefSkip(l) |
+                &Cigar::Del(l)  => {
+                    rpos += l;
+                    c += 1;
+                },
+                &Cigar::HardClip(_) => return None,
+                &Cigar::Back(_) => panic!("Back: Unsupported Cigar operation (not in SAMv1 spec)."),
+                &Cigar::Pad(_) => panic!("Pad: Unsupported Cigar operation (unclear SAMv1 spec).")
             }
         }
-        panic!("read does not enclose SNV");
+        // if no SNV evidence is found before passing the vpos, must be
+        return None;
```

Comment:
There was the following major problem: It accepted `Del`s and `RefSkip`s before a potential `SoftClip` (something that conflicts with the SAMv1 spec) or before any reference match appearing in `seq` (which doesn't make any sense). Thus we first have to find the first Cigar operation that refers to `record.seq()[0]` at `record.pos()`, which is any operation consuming `record.seq()` positions. Only once we have this starting point, can we trust and adjust for following `Del`s and `RefSkip`s by increasing rpos and only then can we take a HardClip as a read end and return `None`. Theoretically, if we wanted to really enforce the SAMv1 spec, we would also have to check that no operation comes between a `SoftClip` and the respective read end or `HardClip` operation... 

Also, if you feel like comparing this to the Cigar string parsing logic of htslib, have a look at resolve_cigar2 in sam.c on github.com/samtools/htslib/sam.c (<https://github.com/samtools/htslib/blob/develop/sam.c#L1861>). ;)

In addition, there were these minor problems:
* the `SoftClip` operation wasn't accounted for, previously
* the `Back` operation doesn't exist in the SAMv1 spec, why is it in `rust-htslib`?
* the `Pad` operation doesn't have a clear explanation in the SAMv1 spec, so I wouldn't use it, as I wouldn't know how
* the loop couldn't return `None`, if the reference base of the SNV was actually deleted or skipped in the record
* qpos wasn't adjusted for the actual position within the current Cigar operation, but was always the first position of this Cigar operation

I hope, I have recapped all the things that have changed here. But I wouldn't be surprised to find further bugs / corner cases here in the future, especially with data from the wild...